src/Shipments/PriceCalculator.php

```diff
@@ -3,155 +3,134 @@
-            && $group->getStepSize()
-            && $group->getStepPrice()) {
-            $price +=
-                ceil(($shipment->getWeight() - $group->getWeightMax()) / $group->getStepSize())
-                * $group->getStepPrice();
-        } elseif ($shipment->getWeight() > $group->getWeightMax() || $shipment->getWeight() < $group->getWeightMin()) {
-            throw new CalculationException(
-                'Cannot calculate a price for given shipment; contract did not contain weight prices for: ' . $shipment->getWeight()
-            );
+            $price += $prices[$option->getId()];
         }
 
         return (int)$price;
     }
 
     /**
-     * Calculate the price based on the selected options for given shipment.
-     * Optionally a contract can be supplied for the price calculations. If no
-     * contract is given, the contract on the shipment is used.
-     *
-     * @param ShipmentInterface             $shipment
-     * @param ServiceContractInterface|null $contract
-     * @return int
+     * @param ShipmentInterface $shipment
+     * @throws InvalidResourceException
      */
-    public function calculateOptionsPrice(ShipmentInterface $shipment, ServiceContractInterface $contract = null)
+    private function validateShipment(ShipmentInterface $shipment)
     {
-        if ($contract === null) {
-            $contract = $shipment->getServiceContract();
+        if ($shipment->getContract() === null) {
+            throw new InvalidResourceException(
+                'Cannot calculate shipment price without a set contract.'
+            );
         }
-
-        if ($contract === null) {
-            throw new CalculationException('Cannot calculate a price for given shipment without a contract');
+        if ($shipment->getService() === null) {
+            throw new InvalidResourceException(
+                'Cannot calculate shipment price without a set service.'
+            );
         }
+    }
 
-        $price = 0;
+    /**
+     * @param ShipmentInterface $shipment
+     * @return ServiceRateInterface
+     */
+    private function determineServiceRateForShipment(ShipmentInterface $shipment)
+    {
+        $this->validateShipment($shipment);
 
-        $optionPrices = [];
-        foreach ($contract->getServiceOptionPrices() as $option) {
-            $optionPrices[$option->getServiceOption()->getId()] = $option->getPrice();
-        }
+        $serviceRates = $shipment->getService()->getServiceRates([
+            'contract' => $shipment->getContract(),
+            'weight'   => $shipment->getPhysicalProperties()->getWeight(),
+        ]);
 
-        foreach ($shipment->getServiceOptions() as $option) {
-            if (!isset($optionPrices[$option->getId()])) {
-                throw new CalculationException(
-                    'Cannot calculate a price for given shipment; invalid option: ' . $option->getId()
-                );
-            }
+        $serviceRate = reset($serviceRates);
```

Comment:
To determine the correct service rate you should also check that the service rate at least has all the service options set on the shipment.